scripts/clean_osm_data.py

```diff
@@ -279,279 +288,320 @@ def finalize_lines_type(df_lines):
+        _parse_cables_to_len, axis=1
+    )
+
+    # indirect matching exploiting the total numeric value in cables
+    # the minimum requirement of cables by frequency value is calculated
+    # then using the total numeric cables number, the values are scaled proportionally
+    df_indirect = df[to_fill_indirect]
+
+    basic_cables = (
+        df_indirect["tag_frequency"]
+        .map(lambda x: [_basic_cables(v) for v in x])
+        .rename("basic_cables")
+    )
+    min_cables = basic_cables.map(sum).rename("min_cables")
+    multiplier = (
+        pd.to_numeric(df_indirect["cables"], errors="coerce") / min_cables
+    ).rename("multiplier")
+    filled_values = pd.concat([basic_cables, multiplier], axis=1).apply(
+        lambda x: ";".join([str(x["multiplier"] * v) for v in x["basic_cables"]]),
+        axis=1,
+    )
+    df.loc[filled_values.index, "circuits"] = filled_values
+
+    # otherwise assume a circuit per element
+    df_fill_default = df[to_fill_default]
+    df.loc[df_fill_default.index, "circuits"] = len_f.loc[df_fill_default.index].map(
+        lambda x: ";".join(["1"] * x)
+    )
+
+    # explode column
+    df["circuits"] = df["circuits"].str.split(";")
+
+    return df
+
+
+def explode_columns(df, cols):
+    # check if all row elements are list
+    is_all_list = df[cols].applymap(lambda x: isinstance(x, list)).all(axis=1)
+    if not is_all_list.all():
+        df_nonlist = df[~is_all_list]
+        logger.warning(
+            f"Unexpected non-list values in dataframe; dropping rows. Needed fix for entries:\n{df_nonlist}"
         )
-        unknown_circuits_tags = set(
-            df_all_lines.loc[not_resolved_circuits]["circuits"].values
+        df.drop(df_nonlist.index, inplace=True)
+
+    # check if errors in the columns
+    nunique_values = df[cols].applymap(len).nunique(axis=1)
+    df_nunique = df[nunique_values != 1]
+    if not df_nunique.empty:
+        logger.warning(
+            f"Improper explosion of dataframe entries; dropping rows. Needed fix for entries:\n{df_nunique}"
         )
-        if any(not_resolved_circuits):
-            df_all_lines.drop(df_all_lines[not_resolved_circuits].index, inplace=True)
-            logger.warning(
-                f"The lines with an unexpected circuits tag value in {unknown_circuits_tags} will be dropped."
-            )
+        df.drop(df_nunique.index, inplace=True)
 
-        df_all_lines["circuits"] = df_all_lines["circuits"].astype(int)
+    df = df.explode(cols, ignore_index=True)
 
-    # drop column if exist
-    if "cables" in df_all_lines:
-        df_all_lines.drop(columns="cables", inplace=True)
+    return df
 
-    return df_all_lines
+
+def integrate_lines_df(df_all_lines, distance_crs):
+    """
+    Function to add underground, under_construction, frequency and circuits
+    """
+    # explode frequency and columns
+    df = pd.DataFrame(df_all_lines)
+
+    # preliminary raw parsing of raw columns
+    parse_frequency(df)
```

Comment:
as mentioned various commetns above, check about if renamig to clean_frequency etc...

Reply:
Done